### PR TITLE
PXB-2722 password is written into xtrabackup_info

### PR DESCRIPTION
**Pull request by rahulmalik87**
https://jira.percona.com/browse/PXB-2722

Analysis:
password passed with -p option is written into backup tool_command in xtrabackup_info

Fix:
mask password before writting into xtrabackup_info

### CODE REVIEW

**Comment by altmannmarcelo** [comment]
@rahulmalik87 I see the test covering the history table, however, the commit mentions it has been written to xtrabackup_info file. Can you add that to the test case?

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 1]
```diff
@@ -2088,7 +2088,10 @@ char *make_argv(char *buf, size_t len, int argc, char **argv)
 		if (strncmp(*argv, "--password", strlen("--password")) == 0) {
 			arg = "--password=...";
 		}
-		if (strncmp(*argv, "--encrypt-key",
+                if (strncmp(*argv, "-p", strlen("-p")) == 0) {
+                  arg = "-p...";
+                }
+                if (strncmp(*argv, "--encrypt-key",
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: plz check alignment. Also check new parameters that are passing cloud info credentials.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+mysql -e "GRANT ALL ON *.* TO pxb@localhost "
+mkdir -p $topdir/backup
+
+# pass either --password or -p
+function grep_for_password() {
+  password=$1$password_string
+  $XB_BIN $XB_ARGS --backup -u pxb $password \
+  --encrypt=AES256 --encrypt-key=r______$password_string --transition-key=$password_string \
+  --target-dir=$topdir/backup \
+  --debug-sync="data_copy_thread_func"  2>&1 | tee $topdir/pxb.log &
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: Do we need debug_sync here? This will make the test not run in half of platforms. 
Perhaps use some approach that will block FTWRL for non-innodb tables, like have SELECT *, SLEEP(6000) FOR UPDATE or something like that.

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  while [ ! -r "$pid_file" ]
+  do
+      sleep 1
+      i=$((i+1))
+      echo "Waited $i seconds for $pid_file to be created"
+  done
+  xb_pid=`cat $pid_file`
+
+  vlog "check for $password_string string in ps "
+  if ps -ef | grep '$password_string' | grep -v grep
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: not that this would be a big issue, but I think we have a potential for other tests running in parallel to interfere, like history_on_server.sh which uses the same password and being executed on a stage prior to stripping the password out. I would indicate use to limit `ps` for the current xtrabackup PID.

Also, I don't think we have any other test that uses `ps`. If I am not mistaken, `ps` is provided by procps library which is not installed by default on Jenkins images.

I have a JIRA ticket to install in as it's needed on PXB-1818

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 3 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: agree @altmannmarcelo it would be better to include xtrabackup.

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 1 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: thank you @satya-bodapati for cloud suggestion. would be adding xbcloud test-case as part of PXB-2723.

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 2 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: fixed

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 3 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: fixed

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 4]
```diff
@@ -11,37 +11,26 @@ vlog "Preparing server"
...
 mysql -e "GRANT ALL ON *.* TO pxb@localhost "
+mysql -e "CREATE TABLE t1(i int) engine=MYISAM" test
+mysql -e "INSERT INTO t1 values(1)" test
 mkdir -p $topdir/backup
 
 #pass option like password -p to be run with 
 function grep_in_datadir_ps() {
-	additional_options=$1
+  additional_options=$1
+  mysql -e "SELECT SLEEP(15) from t1 for update" test &
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: I would make this a higher sleep, like 6000 and capture it's PID and kill it at the end of the backup. Jenkins can be unpredictable on this sense and 15 seconds might no be enough.

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 5]
```diff
@@ -11,37 +11,26 @@ vlog "Preparing server"
...
-  while [ ! -r "$pid_file" ]
-  do
-      sleep 1
-      i=$((i+1))
-      echo "Waited $i seconds for $pid_file to be created"
-  done
-  xb_pid=`cat $pid_file`
-
+  #sleep sufficient time to let backup run
+  sleep 5
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: Same here, sleep won't be reliable. I would rather either grep the output file and check for a string that indicates the process has started, or keep running a kill -0 on pxb PID.

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 6]
```diff
@@ -11,37 +11,26 @@ vlog "Preparing server"
...
   vlog "check for $password_string string in ps "
-  if ps -ef | grep '$password_string' | grep -v grep
+  if ps -ef | grep "xtrabackup.*$password_string" | grep -v grep
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: This will not isolate the ps to only capture xtrabackup been execute on this test. You should run instead

```
ps -ef | grep "${job_pid}"  | grep "${password_string}" | grep -v grep
```

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 7]
```diff
@@ -11,37 +11,26 @@ vlog "Preparing server"
...
   then
      die "found $password_string string in ps output"
   fi
 
-  # Resume xtrabackup
-  vlog "Resuming xtrabackup"
-  kill -SIGCONT $xb_pid
-
   run_cmd wait $job_pid
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh:54]: here we should make sure we kill the SELECT SLEEP(X)

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 4 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: added

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 6 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: fixed.

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 7 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh:54]: fixed

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 8]
```diff
@@ -15,22 +18,31 @@ mysql -e "CREATE TABLE t1(i int) engine=MYISAM" test
...
   2>&1 | tee $topdir/pxb.log &
   job_pid=$!
-  #sleep sufficient time to let backup run
-  sleep 5
-  vlog "check for $password_string string in ps "
-  if ps -ef | grep "xtrabackup.*$password_string" | grep -v grep
+
+  wait_for_process_to_start $job_pid
+  #sleep is to ensure password is removed by xtrabackup
+  sleep 3
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: Does this mean there is window that xtrabackup shows password?

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 9]
```diff
@@ -2088,7 +2088,10 @@ char *make_argv(char *buf, size_t len, int argc, char **argv)
 		if (strncmp(*argv, "--password", strlen("--password")) == 0) {
 			arg = "--password=...";
 		}
-		if (strncmp(*argv, "--encrypt-key",
+		if (strncmp(*argv, "-p", strlen("-p")) == 0) {
+			arg = "-p...";
+		}
+                if (strncmp(*argv, "--encrypt-key",
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: please indent this properly

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 8 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: yes there is window when xtrabackup shows password.  PXB removes the argument passed to script.

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: afair, this is done as soon as soon the argument processing is done. Shouldn't it be less than fraction of second? does this test fail without the sleep?

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 10]
```diff
@@ -1150,5 +1150,27 @@ function backup_and_restore() {
...
 
+### kill all running queries
+function kill_all_queries()
+{
+  run_cmd $MYSQL $MYSQL_ARGS --force --batch  test <<EOF
+  select concat('KILL ',id,';') from information_schema.processlist
+  where user='root' and time > 1 into outfile '$MYSQLD_TMPDIR/killall.sql';
+  source $MYSQLD_TMPDIR/killall.sql;
+EOF
+  rm -f $MYSQLD_TMPDIR/killall.sql
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/common.sh]: isn't it better if we just kill the pid of the MySQL client process? I think such function already exists?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 10 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: killing pid is not help full. the process still exists can be seen with show processlist.

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 8 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: yes this is done after argument processing. yes test fails randomly without sleep.

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 10 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/inc/common.sh]: well, we have to use wait $pid like it is done in other tests.
It looks like this code is duplicated from kill_long_selects.sh. Please move to common place

It is not clear to me if "bg_kill_ok" is used, why do we do "kill_all_queries". Please check with marce

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 8 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: can you please check (by code or test) if the same behaviour applies for "mysql" client tool as well?

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: yes. same thing happen with mysql client as well.

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: ok

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 9 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: what is the wrong indentation?

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 10 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: these is code from 2013. do you have any idea @altmannmarcelo   but I have removed duplicated code from kill_long_selects.sh.  let me know if it is fine?

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 11]
```diff
@@ -1150,5 +1150,27 @@ function backup_and_restore() {
...
+
+function wait_for_process_to_start() {
+  local pid=$1
+  vlog "wait for $pid to start "
+  while  ! kill -0 $pid >/dev/null 2>&1
+  do
+    vlog "waiting for pid $pid to start"
+    sleep 1
+  done
+}
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/common.sh]: I don't think this loop is necessary. Check existing kill_long_selects.sh which use "wait $pid". There is an existing function. Please use that and move to .inc file

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 11 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: sorry I didn't understand . can you please tell me exact function?

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/inc/common.sh]: bg_wait_ok

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: IIUC bg_wait_ok is just wait untill the process is finish. but it is the background process and we wait  until it is started.

Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/inc/common.sh]: Hmm, if you have pid, then process is already started right? why do u need this hack?

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 10 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/test/inc/common.sh]: @rahulmalik87 same as on the azure PR. I think that a kill -9 on mysql client pid will do the trick, have you tested it ?

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+password_string=secretpwd
+
+MYSQLD_EXTRA_MY_CNF_OPTS="
+secure-file-priv=$TEST_VAR_ROOT
+"
+vlog "Preparing server"
+start_server
+mysql -e "CREATE USER pxb@localhost IDENTIFIED BY '$password_string'"
+mysql -e "GRANT ALL ON *.* TO pxb@localhost "
+mysql -e "CREATE TABLE t1(i int) engine=MYISAM" test
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh:17]: explain why we need myisam engine table. What if someone tries to change to innodb.
Also we might need a relook when myisam gets removed

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+mysql -e "INSERT INTO t1 values(1)" test
+mkdir -p $topdir/backup
+
+#pass option like password -p to be run with
+
+function grep_in_datadir_ps() {
+
+  additional_options=$1
+  mysql -e "SELECT SLEEP(1000) from t1 for update " test &
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh:29]: why is this necessary? how does it help.  Please add comments

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 14]
```diff
@@ -1150,5 +1150,29 @@ function backup_and_restore() {
...
+  local condition=$1
+  run_cmd $MYSQL $MYSQL_ARGS --force --batch  test <<EOF
+  select concat('KILL ',id,';') from information_schema.processlist
+  where $condition into outfile '$MYSQLD_TMPDIR/killcondition.sql';
+  source $MYSQLD_TMPDIR/killcondition.sql;
+EOF
+ rm -f $MYSQLD_TMPDIR/killcondition.sql
+}
+
+function wait_for_file_to_generated() {
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/common.sh:1169]: add a comment here too, how this wait for the file to be generated is useful?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
mostly looks good rahul. Some documentation requests. Now that the testcase is changed, I want to see if this passes on both debug and release builds.

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 15]
```diff
@@ -1150,5 +1150,33 @@ function backup_and_restore() {
...
+## example kill_condition_queries "time > 0 & INFO like '%SLEEP%'"
+function kill_condition_queries()
+{
+  local condition=$1
+  run_cmd $MYSQL $MYSQL_ARGS --force --batch  test <<EOF
+  select concat('KILL ',id,';') from information_schema.processlist
+  where $condition into outfile '$MYSQLD_TMPDIR/killcondition.sql';
+  source $MYSQLD_TMPDIR/killcondition.sql;
+EOF
+ rm -f $MYSQLD_TMPDIR/killcondition.sql
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/common.sh]: EOF and rm -rf are wrongly aligned

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 16]
```diff
@@ -1150,5 +1150,33 @@ function backup_and_restore() {
   verify_db_state $db
 }
 
+### kill all running queries accept the condition argument
+## example kill_condition_queries "time > 0 & INFO like '%SLEEP%'"
+function kill_condition_queries()
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/common.sh]: rename to kill_query_pattern() , you can tell the pattern is passed as argument or WHERE condition

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 17]
```diff
@@ -1150,5 +1150,33 @@ function backup_and_restore() {
...
+  run_cmd $MYSQL $MYSQL_ARGS --force --batch  test <<EOF
+  select concat('KILL ',id,';') from information_schema.processlist
+  where $condition into outfile '$MYSQLD_TMPDIR/killcondition.sql';
+  source $MYSQLD_TMPDIR/killcondition.sql;
+EOF
+ rm -f $MYSQLD_TMPDIR/killcondition.sql
+}
+
+### wait for file to be generated.
+### start some process and wait untill that proceses generate the file
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/common.sh]: this can be used to check if  backup has started or not

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+mysql -e "CREATE TABLE t1(i int) engine=MYISAM" test
+mysql -e "INSERT INTO t1 values(1)" test
+mkdir -p $topdir/backup
+
+#pass option like password -p to be run with
+
+function grep_in_datadir_ps() {
+
+  additional_options=$1
+  ### This SLEEP with update on MyISAM will let backup wait untill this query is killed. PXB waits to acquire FLUSH TABLES on MyISAM tables
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: The running update on myisam table will hold access to the table. PXB will do FLUSH TABLES on MyISAM tables which requires access to the table and waits for the running update on MyISAM to finish.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 15 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: EOF should come without any spaces. rm -rf is aligned to 2 spaces it it inside function.

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 16 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: fixed

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 17 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/inc/common.sh]: fixed

File: storage/innobase/xtrabackup/test/t/check_for_passwords.sh (hunk 18 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/test/t/check_for_passwords.sh]: fixed

**Comment by rahulmalik87** [comment]
./run.sh -t t/check_for_passwords.sh
Using /home/rahul.malik/MySQL/build/x7/xtrabackup-test/var as test root
Detecting server version...
Running against MySQL 5.7.37-debug-log (InnoDB 5.7.37)
Using 'xtrabackup' as xtrabackup binary
Autodetected number of cores: 24
Limiting to 16 to avoid excessive resource consumption
Using 16 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
check_for_passwords                      w1	[passed]    27
==============================================================================
Spent 27 of 34 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

See results/ for detailed output

**Review by satya-bodapati — APPROVED** [review]

**Comment by satya-bodapati** [comment]
@rahulmalik87 please squash

**Review by altmannmarcelo — APPROVED** [review]
LGTM. Thanks for addressing all the comments.

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 10 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/test/inc/common.sh]: @rahulmalik87 please test. kill -9 should do imho

**Review by satya-bodapati — APPROVED** [review]
